### ISSUE TITLE
generated code can't compile - type conversion error param/double in indexed expr

### ISSUE THREAD

**Issue by mitzimorris** [stan-dev/stan#2505]
#### Summary:
Reported on Stan forums:
http://discourse.mc-stan.org/t/mysterious-compilation-error-in-mixed-effects-glm-cannot-convert-const-stan-math-var-to-double-in-return-return-u/3738

#### Description:
see discussion in forum - parser fails to detect an expression which
contains references to vars; generated code fails to compile.

#### Reproducible Steps:
TBD
If you have a reproducible example, please include it.

#### Current Output:
see report on forums

#### Expected Output:
Model should compile and run.

#### Additional Information:
Provide any additional information here.

#### Current Version:
v2.17.1

**Comment by mitzimorris** [comment]
tested on current develop branch, edited version of model included on forums is:

```
data {
    int<lower=0> P;
    int<lower=1> D;
}
parameters {
    vector[P] bCov;
    vector[D] alpha;
    vector[D] bRandom;
    cov_matrix[2] Sigma;
}
model {
    for (d in 1:D) { 
        [alpha[d], bRandom[d]]' ~ multi_normal([0,0]',Sigma);
    }
}
```

this compiles.

checking generated code for expression `[alpha[d], bRandom[d]]'`

```
transpose(stan::math::to_row_vector(stan::math::array_builder<local_scalar_t__ >()
.add(get_base1(alpha,d,"alpha",1)).add(get_base1(bRandom,d,"bRandom",1))
.array()))
```

(added linebreaks for readability)  - generated code appropriate for param vars.

**Comment by mitzimorris** [comment]
verified that it is broken in the current release -  2.17.3
used   RStan Version 2.17.3, GitRev: 2e1f913d3ca3,

generated code at line 267:
```
 transpose(stan::math::to_row_vector(stan::math::array_builder<double()
.add(get_base1(alpha,d,"alpha",1)).add(get_base1(bRandom,d,"bRandom",1))
.array())
```

which triggers a compile-time err.

**Comment by mitzimorris** [comment]
at a minimum, we should add a simple version of the above to the good test-models repository.

**Comment by bob-carpenter** [comment]
@mitzimorris  Please assign labels (is this a bug or feature is most important) and then a deadline and ideally assign at least one person to it if the deadline is sooner than "future".

What should the generated code look like?  Or is this a problem in the math lib?

**Comment by mitzimorris** [comment]
closing now - it is fixed in release 2.18.1
